src/System.Linq/tests/UnionTests.cs

```diff
@@ -413,5 +414,15 @@ public void HashSetWithBuiltInComparer_HashSetContainsNotUsed()
             Assert.Equal(new[] { "A", "a" }, input2.Union(input1, EqualityComparer<string>.Default));
             Assert.Equal(new[] { "A" }, input2.Union(input1, StringComparer.OrdinalIgnoreCase));
         }
+
+        [Fact]
+        public void UnionFollowedBySelectOnEmptyEnumerableInvokesDispose()
+        {
+            var enum1 = new DisposeTrackingEnumerable<int>();
+            var enum2 = new DisposeTrackingEnumerable<int>();
+            enum1.Union(enum2).Select(x => x).ToList();
```

Comment:
@gabeluci pointed out in his comment (https://github.com/dotnet/corefx/issues/40384#issuecomment-522198937) that this only occurs when using Select() after Union, as otherwise it ends up in [UnionWith](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Set.cs#L202).

Reply:
Upon further testing it seems like there are some other issues as well.

This works:
```csharp
            var enum1 = new DisposeTrackingEnumerable<int>();
            var enum2 = new DisposeTrackingEnumerable<int>();

            enum1.Union(enum2).ToList();

            Assert.True(enum1.EnumeratorDisposed);
            Assert.True(enum2.EnumeratorDisposed);
```

Whereas this doesn't

```csharp
            var enum1 = new DisposeTrackingEnumerable<int>();
            var enum2 = new DisposeTrackingEnumerable<int>();

            foreach (int value in enum1.Union(enum2)) { }

            Assert.True(enum1.EnumeratorDisposed);
            Assert.True(enum2.EnumeratorDisposed);
```

Reply:
Ah my bad, I misread your comment. I'll make the proposed changes!